### PR TITLE
hclust() refactoring + Ward method

### LINKED ISSUE #109: HClust initializes the Nearest Neighbor Array to 0-index leading to failure.
At places the code initializes the nearest neighbor array to 0-index leading to failure. This is particularly for the :single method when one of data points is not linked or a group of data points are not linked.

### PR DESCRIPTION
**Pull request by alyst**
It started as a one line fix to reduce the number of allocations.
~~Actually, despite the big diff, the code of @davidavdav is quite efficient.~~ Now it's much faster, since it uses Lance-Williams approach for updating inter-cluster distances after clusters merging.

### Notable changes:
* Much faster `:average` and `:complete`
* `method` renamed to `linkage` (i.e. _cluster linkage_). It's less confusing (otherwise one may think it's about the actual algorithm to build the tree), also this is how it's called in e.g. Python's sklearn.
* Ward cluster linkage (`:ward` (=`Ward.D2` of R) and `:ward_presquared` (=`Ward.D1` of R)). It's implemented by `hclust_nn_lw()` internal method that accepts `metric::ReducibleMetric` param (_reducible_ Lance-Williams cluster-cluster distance). After the PR is landed it should be fairly straightforward to add the other Lance-Williams metrics (WPGMA, Centroid, Median -- the latter two are not reducible, but `hclust_nn_lw()` should still generate good approximation of the optimal tree)
* `Hclust` struct fields were changed (it carried too much R heritage): `height` => `heights`, `merge` => `merges`, `method` => `linkage`, `labels` field removed (there was no support for it anyway). The old field names are still available (via `getproperty()`) during the deprecation period.
* Internal refactoring to move the common code into separate structs/methods, e.g. internal `HclustMerges` and `HclustTrees` structs
* reduced GC overhead (in particular, by avoiding allocating 1-element vectors for the leaf nodes)
* the cluster numbers returned by `cutree()` match the R version
* `hclust_n3()` and `cutree()` covered by tests
* added rudimentary benchmark scripts
* fixes #109

## Benchmarks (the script is in `benchmark` folder):

| func   | param         | linkage  | time_old   | time_new   | gcratio_old | gcratio_new | allocs_old | allocs_new | memory_old | memory_new | n_old  | n_new |
|--------|---------------|----------|------------|------------|-------------|-------------|------------|------------|------------|------------|--------|-------|
| cutree | (10, 3)       | single   | 1.805 μs   | 11.412 μs  | 0.00%       | 0.00%       | 71         | 15         | 5.67 KiB   | 1.16 KiB   | 10000  | 902   |
| cutree | (100, 10)     | single   | 21.381 μs  | 21.040 μs  | 0.00%       | 0.00%       | 790        | 77         | 75.94 KiB  | 11.36 KiB  | 10000  | 938   |
| cutree | (1000, 100)   | single   | 302.779 μs | 211.556 μs | 0.00%       | 0.00%       | 7836       | 1000       | 1.36 MiB   | 464.11 KiB | 10000  | 499   |
| cutree | (10000, 1000) | single   | 19.788 ms  | 16.917 ms  | 42.36%      | 54.46%      | 78297      | 13037      | 70.10 MiB  | 51.24 MiB  | 225    | 11    |
| hclust | 100           | average  | 713.783 μs | 77.113 μs  | 0.00%       | 0.00%       | 850        | 301        | 77.39 KiB  | 112.53 KiB | 6843   | 10000 |
| hclust | 100           | complete | 927.169 μs | 94.241 μs  | 0.00%       | 0.00%       | 770        | 318        | 69.39 KiB  | 113.42 KiB | 5305   | 10000 |
| hclust | 100           | single   | 267.058 μs | 78.902 μs  | 0.00%       | 0.00%       | 1641       | 74         | 187.78 KiB | 96.53 KiB  | 10000  | 10000 |
| hclust | 10            | complete | 8.634 μs   | 4.754 μs   | 0.00%       | 0.00%       | 97         | 56         | 7.31 KiB   | 4.89 KiB   | 10000  | 10000 |
| hclust | 10            | average  | 8.264 μs   | 4.181 μs   | 0.00%       | 0.00%       | 97         | 52         | 7.34 KiB   | 4.70 KiB   | 10000  | 10000 |
| hclust | 10            | single   | 17.332 μs  | 2.783 μs   | 0.00%       | 0.00%       | 147        | 29         | 8.80 KiB   | 3.31 KiB   | 10000  | 10000 |
| hclust | 10000         | complete | 57.037 s   | 2.520 s    | 0.00%       | 0.35%       | 76724      | 45128      | 7.58 MiB   | 767.69 MiB | 1      | 14    |
| hclust | 10000         | single   | 3.544 s    | 3.104 s    | 4.40%       | 4.35%       | 173901     | 11997      | 976.94 MiB | 895.62 MiB | 2      | 12    |
| hclust | 10000         | average  | 54.697 s   | 2.341 s    | 0.00%       | 0.38%       | 88237      | 43384      | 8.16 MiB   | 767.80 MiB | 1      | 15    |
| hclust | 1000          | complete | 174.312 ms | 13.866 ms  | 0.00%       | 0.00%       | 7285       | 3717       | 711.88 KiB | 8.02 MiB   | 29     | 2586  |
| hclust | 1000          | average  | 136.956 ms | 11.137 ms  | 0.00%       | 0.00%       | 8386       | 3550       | 826.61 KiB | 8.02 MiB   | 36     | 2929  |
| hclust | 1000          | single   | 12.352 ms  | 12.805 ms  | 0.00%       | 0.00%       | 16179      | 911        | 10.46 MiB  | 8.80 MiB   | 395    | 2493  |

Maybe it also helps #93

### CODE REVIEW

**Comment by alyst** [comment]
So I've rebased it and updated to v0.7 plus implemented fixes and tests for #109.
Here's the latest benchmarks:

| func   | param         | method   | time_old   | time_new   | gcratio_old | gcratio_new | allocs_old | allocs_new | memory_old | memory_new | samples_old | samples_new |
|--------|---------------|----------|------------|------------|-------------|-------------|------------|------------|------------|------------|---------------|---------------|
| cutree | (100, 10)     | missing  | 21.381 μs  | 6.337 μs   | 0.00%       | 0.00%       | 790        | 71         | 75.94 KiB  | 11.61 KiB  | 10000         | 10000         |
| cutree | (10000, 1000) | missing  | 19.788 ms  | 44.891 ms  | 42.36%      | 78.65%      | 78297      | 13029      | 70.10 MiB  | 53.58 MiB  | 225           | 2             |
| cutree | (10, 3)       | missing  | 1.805 μs   | 838.000 ns | 0.00%       | 0.00%       | 71         | 15         | 5.67 KiB   | 1.16 KiB   | 10000         | 10000         |
| cutree | (1000, 100)   | missing  | 302.779 μs | 174.909 μs | 0.00%       | 0.00%       | 7836       | 1004       | 1.36 MiB   | 514.67 KiB | 10000         | 255           |
| hclust | 100           | average  | 713.783 μs | 251.955 μs | 0.00%       | 0.00%       | 850        | 203        | 77.39 KiB  | 33.02 KiB  | 6843          | 10000         |
| hclust | 100           | single   | 267.058 μs | 74.105 μs  | 0.00%       | 0.00%       | 1641       | 73         | 187.78 KiB | 96.66 KiB  | 10000         | 10000         |
| hclust | 100           | complete | 927.169 μs | 255.127 μs | 0.00%       | 0.00%       | 770        | 220        | 69.39 KiB  | 33.75 KiB  | 5305          | 10000         |
| hclust | 10            | average  | 8.264 μs   | 5.204 μs   | 0.00%       | 0.00%       | 97         | 44         | 7.34 KiB   | 3.88 KiB   | 10000         | 10000         |
| hclust | 10            | complete | 8.634 μs   | 5.050 μs   | 0.00%       | 0.00%       | 97         | 48         | 7.31 KiB   | 4.11 KiB   | 10000         | 10000         |
| hclust | 10            | single   | 17.332 μs  | 3.354 μs   | 0.00%       | 0.00%       | 147        | 29         | 8.80 KiB   | 3.30 KiB   | 10000         | 10000         |
| hclust | 10000         | complete | 57.037 s   | 10.812 s   | 0.00%       | 0.00%       | 76724      | 35329      | 7.58 MiB   | 4.74 MiB   | 1             | 1             |
| hclust | 10000         | average  | 54.697 s   | 11.747 s   | 0.00%       | 0.00%       | 88237      | 33537      | 8.16 MiB   | 4.75 MiB   | 1             | 1             |
| hclust | 10000         | single   | 3.544 s    | 2.755 s    | 4.40%       | 4.99%       | 173901     | 12192      | 976.94 MiB | 897.92 MiB | 2             | 2             |
| hclust | 1000          | complete | 174.312 ms | 53.446 ms  | 0.00%       | 0.00%       | 7285       | 2740       | 711.88 KiB | 391.44 KiB | 29            | 85            |
| hclust | 1000          | average  | 136.956 ms | 56.907 ms  | 0.00%       | 0.00%       | 8386       | 2597       | 826.61 KiB | 391.72 KiB | 36            | 81            |
| hclust | 1000          | single   | 12.352 ms  | 11.281 ms  | 0.00%       | 0.00%       | 16179      | 919        | 10.46 MiB  | 8.83 MiB   | 395           | 301           |

I think there was some glitch in `cutree (10000, 1000)`, which would go away in the next try. In the other tests the new code shows improvements.
As before, the results are compared with R's `hclust()` for a few hundred random datasets, so the algorithms should be correct.

It would be nice if somebody reviews.

This PR introduces some helper functions to facilitate the construction of the hierarchical clustering. My plan is to implement Ward clustering (or a generic Lance-Williams one) on top of it in the PR to follow. There was already #94, but it had license issues. The new code will be written from scratch, so it should have no license issues.

**Comment by alyst** [comment]
This PR is already big, so I've added Ward (`:ward` == R's `Ward.D2` and `:ward_presquared` == R's `Ward.D1`) here. Ward and any other Lance-Williams metrics (including `:average` and `:complete`) are supported by the new `hclust_nn_lw()` (in the end it's not much different from `hclust2()`, but much faster since it doesn't recalculate the metric from point distances for every comparison). I've made `hclust_nn_lw()` the default method. It uses more RAM (it has to maintain its copy of distance matrix), but that should be fine.

I've renamed some `Hclust` fields (`height` => `heights` etc) and implemented support for the old names during deprecation period via `getproperty()`.

I would merge it in a few days, if nobody objects. Still happy to see the reviews. :)

The new benchmark (`:single` got 10% slower since the last benchmark, because I made it share some common routines with the `hclust_nn_lw()`  (neighbor search, matrix updates), but it should not matter much in practice).

| func   | param         | linkage  | time_old   | time_new   | gcratio_old | gcratio_new | allocs_old | allocs_new | memory_old | memory_new | n_old  | n_new |
|--------|---------------|----------|------------|------------|-------------|-------------|------------|------------|------------|------------|--------|-------|
| cutree | (10, 3)       | single   | 1.805 μs   | 11.412 μs  | 0.00%       | 0.00%       | 71         | 15         | 5.67 KiB   | 1.16 KiB   | 10000  | 902   |
| cutree | (100, 10)     | single   | 21.381 μs  | 21.040 μs  | 0.00%       | 0.00%       | 790        | 77         | 75.94 KiB  | 11.36 KiB  | 10000  | 938   |
| cutree | (1000, 100)   | single   | 302.779 μs | 211.556 μs | 0.00%       | 0.00%       | 7836       | 1000       | 1.36 MiB   | 464.11 KiB | 10000  | 499   |
| cutree | (10000, 1000) | single   | 19.788 ms  | 16.917 ms  | 42.36%      | 54.46%      | 78297      | 13037      | 70.10 MiB  | 51.24 MiB  | 225    | 11    |
| hclust | 100           | average  | 713.783 μs | 77.113 μs  | 0.00%       | 0.00%       | 850        | 301        | 77.39 KiB  | 112.53 KiB | 6843   | 10000 |
| hclust | 100           | complete | 927.169 μs | 94.241 μs  | 0.00%       | 0.00%       | 770        | 318        | 69.39 KiB  | 113.42 KiB | 5305   | 10000 |
| hclust | 100           | single   | 267.058 μs | 78.902 μs  | 0.00%       | 0.00%       | 1641       | 74         | 187.78 KiB | 96.53 KiB  | 10000  | 10000 |
| hclust | 10            | complete | 8.634 μs   | 4.754 μs   | 0.00%       | 0.00%       | 97         | 56         | 7.31 KiB   | 4.89 KiB   | 10000  | 10000 |
| hclust | 10            | average  | 8.264 μs   | 4.181 μs   | 0.00%       | 0.00%       | 97         | 52         | 7.34 KiB   | 4.70 KiB   | 10000  | 10000 |
| hclust | 10            | single   | 17.332 μs  | 2.783 μs   | 0.00%       | 0.00%       | 147        | 29         | 8.80 KiB   | 3.31 KiB   | 10000  | 10000 |
| hclust | 10000         | complete | 57.037 s   | 2.520 s    | 0.00%       | 0.35%       | 76724      | 45128      | 7.58 MiB   | 767.69 MiB | 1      | 14    |
| hclust | 10000         | single   | 3.544 s    | 3.104 s    | 4.40%       | 4.35%       | 173901     | 11997      | 976.94 MiB | 895.62 MiB | 2      | 12    |
| hclust | 10000         | average  | 54.697 s   | 2.341 s    | 0.00%       | 0.38%       | 88237      | 43384      | 8.16 MiB   | 767.80 MiB | 1      | 15    |
| hclust | 1000          | complete | 174.312 ms | 13.866 ms  | 0.00%       | 0.00%       | 7285       | 3717       | 711.88 KiB | 8.02 MiB   | 29     | 2586  |
| hclust | 1000          | average  | 136.956 ms | 11.137 ms  | 0.00%       | 0.00%       | 8386       | 3550       | 826.61 KiB | 8.02 MiB   | 36     | 2929  |
| hclust | 1000          | single   | 12.352 ms  | 12.805 ms  | 0.00%       | 0.00%       | 16179      | 911        | 10.46 MiB  | 8.80 MiB   | 395    | 2493  |

cc @ararslan (As you were the most active in the repository recently :) )

**Comment by wildart** [comment]
LGTM

**Comment by ararslan** [comment]
> cc @ararslan (As you were the most active in the repository recently :) )

I was? o_O Regardless, this looks good to me.

**Comment by davidavdav** [comment]
Thanks, @alyst for all the work!
